### ISSUE TITLE
What is the philosophy with numbers?

### ISSUE THREAD

**Issue by michaelgwelch** [gregsdennis/json-everything#584]
### Documentation

- [X] I have consulted the [documentation](https://docs.json-everything.net), and my question isn't answered there.

### Nuget Package

JsonSchema.Net

### Package Version

5.4.2

### How can I help?  Please provide as much context as possible.

Previously I was hand coding all of the JSON Schemas I need using `System.Test.Json` with `Utf8JsonWriter`.

What a joy to discover JsonSchema.Net and a better way to do things. However, I may have slammed into a brick wall when I went to use the `builder.ExclusiveMinimum` extension method.

It seems that this method as well as all the methods that take numbers take (and store) decimal values rather than allowing one to specify `float`, or `double` or `decimal`.

Searching for some help on this I even found this comment on a closed Issue: https://github.com/gregsdennis/json-everything/issues/488#issuecomment-1646265100

Basically it states that `decimal` is used as the backing store and due to this the range is restricted to `decimal`.

Am I missing something or does this just mean one can't use `double` values outside the `decimal` range?

I'm thinking that perhaps I could just create my own versions of `minimum`, `maximum`, `exclusiveMinimum` and `exclusiveMaximum` keywords, along with the necessary extension methods. (I'm assuming this would work, as we aren't using the library to deserialize. So theoretically I wouldn't need to register these keywords which I assume I wouldn't be able to do as they use the same keyword strings as the built-in ones).

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gregsdennis** [comment]
Using `decimal` was a design decision because I feel that the higher precision over `double` was more important than the range.

You can certainly create your own keywords to handle `double` if you want, but I'm not sure if they'd be picked up.  Registering them would overwrite the registrations for the default implementations and allow yours to be deserialized.  But evaluation, by default, filters out non-vocabulary keywords, and the vocabs (see `Vocabularies` class) are hard-coded with the included implementations.  You _might_ be able to get it to work if you set the `ProcessCustomKeywords` option, but I'm not sure.

More reading: https://blog.json-everything.net/posts/numbers-are-numbers-not-strings/

**Comment by michaelgwelch** [comment]
@gregsdennis Is this a decision that may be enhanced in the future? Or for now it's not even on your radar?

**Comment by gregsdennis** [comment]
The philosophy isn't going to change; in my experience `decimal` in generally more useful.  But I might be open to including `double` versions of the keywords along with a "switching" method on the keyword registry.  Something like:

```c#
SchemaKeywordRegistry.EnableDoubleSupport();
```

which would swap out all the necessary registrations to use the `double` versions.  I'll need to play with it to see if it works the way I think it would.

**Comment by gregsdennis** [comment]
After building this up a bit, I don't think that I'm going to include this in the main lib.  I might at some point publish an extension, but I'd want to see more demand before I do.

In the meantime, you're welcome to have this.
[DoubleSupport.zip](https://github.com/gregsdennis/json-everything/files/13678234/DoubleSupport.zip)

Just add it to your solution, and it'll just work.

Here's using it:

```c#
DoubleSupport.Enable();

var schema = new JsonSchemaBuilder()
	.Minimum(4.0);
var instance = 3;

var result = schema.Evaluate(instance);
```

**Comment by michaelgwelch** [comment]
Wow! Thanks @gregsdennis 

Really didn't expect anything like this.

**Comment by michaelgwelch** [comment]
With your zip file providing a template on how to approach this problem, I've been able to write a proof of concept that allows for `int`, `uint`, `float` or `double` values (we have no use for `decimal` in our application so I didn't add support for that, but it wouldn't be hard).

The key to my solution is creating a `Number` variant type and then designing some sort of algorithm for deciding which type is best (for us) when parsing a schema (which is truthfully is not a very common use case for us).

I plan to put my code in a repo soon and will share it a link with you.
